ENH: Add dropna in groupby to allow NaN in keys

charlesdong1991 opened the pull request:
- [x] closes #3729 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Note that this PR will *NOT* fix the issue for `pivot_table` for now, the reason is that there is already an argument called `dropna` in `pivot_table` and it has slightly different meaning, currently it means: `Do not include columns whose entries are all NaN`.

I would propose a change in the follow-up PR for this since this is an api change: change the name of current `dropna` to `drop_all_na` maybe? and then add `dropna` to it and it is aligned with the `dropna` in groupby.


Summary:
After this PR, it will optional to inlcude NaN in group keys, e.g. below, and i also add example in docstring as well:
```python
a = [['a', 'b', 12, 12, 12], ['a', None, 12.3, 233., 12], ['b', 'a', 123.23, 123, 1], ['a', 'b', 1, 1, 1.]]
df = pd.DataFrame(a, columns=['a', 'b', 'c', 'd', 'e'])

df.groupby(by=['a', 'b']).sum()
```
will get
![Screen Shot 2020-01-01 at 11 01 16 AM](https://user-images.githubusercontent.com/9269816/71640222-19e68900-2c86-11ea-8749-c223729eec99.png)

with `dropna=False`,
```python
df.groupby(by=['a', 'b'], dropna=False).sum()
```
![Screen Shot 2020-01-01 at 11 01 23 AM](https://user-images.githubusercontent.com/9269816/71640225-2ff44980-2c86-11ea-9c66-a18147179c12.png)

For Series, it is the same:
```python
s = pd.Series([1, 2, 3, 3], index=["a", 'a', 'b', np.nan])

s.groupby(level=0).sum()
s.groupby(level=0, dropna=False).sum()
```
![Screen Shot 2020-01-01 at 11 01 32 AM](https://user-images.githubusercontent.com/9269816/71640228-4d291800-2c86-11ea-9d66-4a4be1a7347d.png)

charlesdong1991 committed:
remove \n from docstring

charlesdong1991 committed:
fix conflicts

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master'

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into fix_issue_3729

charlesdong1991 committed:
fix issue 3729

charlesdong1991 committed:
fix conflicts

charlesdong1991 committed:
not check type

charlesdong1991 committed:
Add groupby test for Series

charlesdong1991 committed:
Add whatsnew note

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -549,6 +549,7 @@ Other API changes
   Supplying anything else than ``how`` to ``**kwargs`` raised a ``TypeError`` previously (:issue:`29388`)
 - When testing pandas, the new minimum required version of pytest is 5.0.1 (:issue:`29664`)
 - :meth:`Series.str.__iter__` was deprecated and will be removed in future releases (:issue:`28277`).
+- :meth:`DataFrame.groupby` and :meth:`Series.groupby` have gained ``dropna`` argument in order to allow ``NaN`` values in group keys (:issue:`3729`)
```
would need a subsection, this is a major new feature

charlesdong1991 replied:
added!

charlesdong1991 replied:
And i also updated on top in description.

jreback commented on pandas/core/algorithms.py:
```diff
@@ -596,7 +596,11 @@ def _factorize_array(
+    sort: bool = False,
+    na_sentinel: int = -1,
+    size_hint: Optional[int] = None,
+    dropna: Optional[bool] = None,
```
shouldn't this be just bool?

charlesdong1991 replied:
changed to `bool`

jreback commented on pandas/core/algorithms.py:
```diff
@@ -630,6 +634,9 @@ def factorize(
             uniques, codes, na_sentinel=na_sentinel, assume_unique=True, verify=False
         )
+    if dropna is False and (codes == na_sentinel).any():
+        uniques = np.append(uniques, [np.nan])
```
ideally we push this down to cython, but ok here

jreback commented on pandas/core/algorithms.py:
```diff
@@ -630,6 +634,9 @@ def factorize(
             uniques, codes, na_sentinel=na_sentinel, assume_unique=True, verify=False
         )
+    if dropna is False and (codes == na_sentinel).any():
+        uniques = np.append(uniques, [np.nan])
```
hmm, this should be a dtype appropriate for the dtype

charlesdong1991 replied:
i think `dtype` is defined below in `_construct_data`? also, if added, for categorical values, will get `ValueError: Categorial categories cannot be null` error.

jreback replied:
right so instead of adding a null to the categories like you are doing, you just add the appropriate -1 entries in the codes which automatically handle the nullness

charlesdong1991 replied:
emm, this has to be a `None` or `NA` value in there. otherwise, output of uniques does not have na value, and also, my python will crash with `FatalError` somehow

jreback commented on pandas/core/generic.py:
```diff
@@ -7355,6 +7356,12 @@ def groupby(
 
             .. versionadded:: 0.23.0
+        dropna : bool or None, default None
+            If None or True, and if group keys contain NaN values, NaN values together
```
WAT? why are you suggesting this tripartite behavior

charlesdong1991 replied:
changed to boolean

jreback reviewed (changes requested):
can you show some examples of using this (in the top of the PR). (ultimately these would need to become doc-examples).

charlesdong1991 committed:
Code change based on JR review

charlesdong1991 committed:
fix conflicts

charlesdong1991 committed:
add forgotten commits

charlesdong1991 committed:
add forgotten commit

charlesdong1991 committed:
Add dropna for series

charlesdong1991 committed:
add doc example for Series

charlesdong1991 committed:
Add level example for series groupby

charlesdong1991 committed:
Add doc example for frame groupby

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -2025,3 +2025,146 @@ def test_groupby_crash_on_nunique(axis):
     tm.assert_frame_equal(result, expected)
+
+
+@pytest.mark.parametrize(
```
can you make a new file, test_groupby_dropna

charlesdong1991 replied:
yes, moved to `test_groupby_dropna`

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -2025,3 +2025,146 @@ def test_groupby_crash_on_nunique(axis):
+        ),
+    ],
+)
+def test_groupby_dropna_multi_index_dataframe_agg(dropna, tuples, outputs):
```
can you add an example that uses NaT (datetime & timedelta)

charlesdong1991 replied:
added! see in `test_groupby_dropna`

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -2025,3 +2025,146 @@ def test_groupby_crash_on_nunique(axis):
+        ["A", None, 12.3, 233.0, 12],
+        ["B", "A", 123.23, 123, 1],
+        ["A", "B", 1, 1, 1.0],
+    ]
```
what if we have NaN in 2 groups? does this work (e.g. different 1st level, but NaN) in 2nd. Also how is nan in first level?

charlesdong1991 replied:
added a lot more different scenarios!

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -191,6 +191,16 @@ method on a :func:`pandas.api.indexers.BaseIndexer` subclass that will generate
+We've added a ``dropna`` keyword to :meth:`DataFrame.groupby` and :meth:`Series.groupby` in order to
+allow ``NaN`` values in group keys. Users can define ``dropna`` to ``False`` if they want to include
+``NaN`` values in groupby keys. The default is set to ``True`` for ``dropna`` to keep backwards
+compatibility (:issue:`3729`)
```
add an example (like you have in the doc-strings)

also add examples in groupby.rst (and point from here)

charlesdong1991 replied:
added in both!

jreback commented on pandas/core/frame.py:
```diff
@@ -5648,6 +5648,41 @@ def update(
 Captive      210.0
 Wild         185.0
+
+We can also choose to include NaN in group keys or not by defining
```
defining -> setting

charlesdong1991 replied:
rephrased

jreback commented on pandas/core/generic.py:
```diff
@@ -7346,6 +7346,12 @@ def clip(
 
             .. versionadded:: 0.23.0
+        dropna : bool, default True
+            If True, and if group keys contain NaN values, NaN values together
```
don't say NaN, say NA values (e.g. can also be NaT or the new NA scalar)

charlesdong1991 replied:
changed!

jreback commented on pandas/core/algorithms.py:
```diff
@@ -596,7 +596,11 @@ def _factorize_array(
+    values,
+    sort: bool = False,
+    na_sentinel: int = -1,
+    size_hint: Optional[int] = None,
```
can you add some independent tests for factorize with dropna

charlesdong1991 replied:
emm, I think i added two test cases already in `test_algos`. Is it what you mean? or you want to have different test cases?

charlesdong1991 committed:
Code change based on JR reviews

charlesdong1991 committed:
add doc

charlesdong1991 committed:
move doc

charlesdong1991 committed:
NaN to NA

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into fix_issue_3729

charlesdong1991 committed:
Fix linting

charlesdong1991 committed:
fix rst issue

charlesdong1991 committed:
fix rst issue

charlesdong1991 commented:
Hi, @jreback 
I addressed all your comments, and merged master, pls let me know if you would like to see other changes.

WillAyd commented on pandas/core/algorithms.py:
```diff
@@ -630,6 +634,9 @@ def factorize(
         uniques, codes = safe_sort(
             uniques, codes, na_sentinel=na_sentinel, assume_unique=True, verify=False
         )
+    if not dropna and (codes == na_sentinel).any():
```
Can you assign `codes == na_sentinel` to a mask variable prior to this? Slightly preferable to re-evaluating this in the branch

charlesdong1991 replied:
done, thanks!

WillAyd commented on pandas/core/algorithms.py:
```diff
@@ -630,6 +634,9 @@ def factorize(
             uniques, codes, na_sentinel=na_sentinel, assume_unique=True, verify=False
         )
+    if not dropna and (codes == na_sentinel).any():
+        uniques = np.append(uniques, [None])
```
These are typically `int64` right? If so just curious if coercing to dtype is going to cause performance regressions

charlesdong1991 replied:
i think `uniques` should be typically `str`/`category`, `codes` are `int`. i will rethink of this issue. thanks!

charlesdong1991 committed:
refactor based on WA review

WillAyd reviewed (approved):
Lgtm

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -191,6 +192,31 @@ method on a :func:`pandas.api.indexers.BaseIndexer` subclass that will generate
+
+    df.groupby(by=["b"], dropna=False).sum()
+
+.. ipython:: python
```
no need to show series

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -191,6 +192,31 @@ method on a :func:`pandas.api.indexers.BaseIndexer` subclass that will generate
+
+.. ipython:: python
+
+    df_list = [[1, 2, 3], [1, None, 4], [2, 1, 3], [1, 2, 2]]
```
make this the same as the docs section; the key is to show both dropna=True (and say its the default) and False

jreback commented on pandas/core/algorithms.py:
```diff
@@ -631,6 +635,11 @@ def factorize(
 
+    code_is_na = codes == na_sentinel
+    if not dropna and code_is_na.any():
+        uniques = np.append(uniques, [None])
```
this is likley changing the dtype; you need the na for this type, use pandas.core.dtypes.cast.na_value_for_dtype.

charlesdong1991 replied:
i do not think so, first of all, if this changed using `cast.na_value_for_dtype`, then `_reconstruct_data` will fail since it cannot cast `NaN` to `integer`, and secondly this `_reconstruct_data` will cast the data to the correct type, so as you see from the output, all data are casted correctly.

WillAyd replied:
Is leveraging the Int64 type an option here? This just seems a little strange in its current state

jreback commented on pandas/core/frame.py:
```diff
@@ -5651,6 +5651,41 @@ def update(
 Wild         185.0
+
+We can also choose to include NaN in group keys or not by setting
+`dropna` parameter:
```
mention the default

jreback commented on pandas/core/groupby/grouper.py:
```diff
@@ -100,7 +100,9 @@ def __new__(cls, *args, **kwargs):
 
-    def __init__(self, key=None, level=None, freq=None, axis=0, sort=False):
+    def __init__(
+        self, key=None, level=None, freq=None, axis=0, sort=False, dropna=True
```
type what you are adding

charlesdong1991 replied:
added

charlesdong1991 replied:
emm, added and got the error below, i think this needs to have typing validation for this `Grouper` class

```
pandas/core/groupby/grouper.py:138: error: Incompatible types in assignment (expression has type "BaseGrouper", variable has type "None")
pandas/core/groupby/grouper.py:138: error: Value of type variable "FrameOrSeries" of "get_grouper" cannot be "None"
pandas/core/groupby/grouper.py:176: error: "None" has no attribute "take"
pandas/core/groupby/grouper.py:204: error: Incompatible types in assignment (expression has type "FrameOrSeries", variable has type "None")
```

jreback replied:
kk can do as a followon

jreback commented on pandas/core/series.py:
```diff
@@ -1621,6 +1621,34 @@ def _set_name(self, name, inplace=False):
 Wild       185.0
 Name: Max Speed, dtype: float64
+
+We can also choose to include NaN in group keys or not by defining
```
mention the default value

charlesdong1991 replied:
mentioned

jreback commented on pandas/tests/groupby/test_groupby_dropna.py:
```diff
@@ -0,0 +1,236 @@
+    mi = pd.MultiIndex.from_tuples(tuples, names=list("ab"))
+    expected = pd.DataFrame(outputs, index=mi)
+
+    tm.assert_frame_equal(grouped, expected, check_index_type=False)
```
why are you setting check_index_type=False?

jreback commented on pandas/tests/groupby/test_groupby_dropna.py:
```diff
@@ -0,0 +1,236 @@
+
+    expected = pd.DataFrame(outputs, index=pd.Index(idx, dtype="object", name="a"))
+
+    tm.assert_frame_equal(grouped, expected, check_index_type=False)
```
why is check_index_type=False?

jreback commented on pandas/tests/groupby/test_groupby_dropna.py:
```diff
@@ -0,0 +1,236 @@
+    mi = pd.MultiIndex.from_tuples(tuples, names=list("ab"))
+    expected = pd.DataFrame(outputs, index=mi)
+
+    tm.assert_frame_equal(grouped, expected, check_index_type=False)
```
why is check_index-type=False, we almost *never* want to do this, likely something is wrong if you are setting this.

charlesdong1991 replied:
emm, this is because of `string` and `np.nan`,  expected will get `string` inferred_type if we use  `mi = pd.MultiIndex.from_tuples(tuples, names=list("ab"))` to define `index` while grouped result will provide `mixed` type. 

i feel the behaviour seems correct? maybe i should change how the `index` is defined? @jreback

charlesdong1991 replied:
ahh, i think i find the issue, i printed the `dtype` and `inferred_dtype` for both, and they are the same, and then i digged a bit into the `util/testing.py`, and found the issue came from `assert_index_equal`, 
https://github.com/pandas-dev/pandas/blob/c99dfea33612f44e97c2365f78c0ca6d5754a1bc/pandas/_testing.py#L667

it's more like a behaviour issue, now if we do below
```python
tuples = [["A", "B"], ["A", np.nan], ["B", "A"]]
mi = pd.MultiIndex.from_tuples(tuples, names=list("ab"))
mi.levels
```
Then `mi.levels` will automatically drop `np.nan` in levels, and that\s why the `inferred_type` of expected is `string`, while grouped result is `mixed` which i think is correct. Because if we directly do:
```python
>>> tuples = [["A", "B"], ["A", np.nan], ["B", "A"]]
>>> mi = pd.MultiIndex.from_tuples(tuples, names=list("ab"))
>>> mi.inferred_type
>>> 'mixed'
```
so what is your thoughts on how to fix it? i would consider it a bug in `assert_index_equal`, since grouped result returns the correct type
@jreback @TomAugspurger

WillAyd replied:
This seems like an issue with the "grouped" result, no?

charlesdong1991 replied:
i think so @WillAyd

jreback commented on pandas/tests/groupby/test_groupby_dropna.py:
```diff
@@ -0,0 +1,236 @@
+    mi = pd.MultiIndex.from_tuples(tuples, names=list("ab"))
+    expected = pd.DataFrame(outputs, index=mi)
+
+    tm.assert_frame_equal(grouped, expected, check_index_type=False)
```
same comment on check_index_type

jreback replied:
why are you not checking the index type? make your expected have the correct type here

jreback commented on pandas/tests/groupby/test_groupby_dropna.py:
```diff
@@ -0,0 +1,236 @@
+        ),
+    ],
+)
+def test_groupby_dropna_datetime_data(na_value1, na_value2, dropna, values, indexes):
```
can you also do an example with Timedelta & Period types as well. (you can just add to the paramaterization)

charlesdong1991 replied:
emm, i think the best/clearest way is to add new tests, so added tests for these two cases.

WillAyd replied:
@charlesdong1991 can you try to parametrize these? Right now there are a lot of tests to read through that I think all do the same thing; parametrization would make things much easier

jreback commented on pandas/tests/test_algos.py:
```diff
@@ -326,6 +326,78 @@ def test_factorize_na_sentinel(self, sort, na_sentinel, data, uniques):
+        codes, uniques = algos.factorize(data, dropna=dropna)
+
+        tm.assert_numpy_array_equal(uniques, expected_uniques)
+        tm.assert_numpy_array_equal(codes, expected_codes, check_dtype=False)
```
why is the dtype not the same?

charlesdong1991 replied:
removed this `check_dtype=False`

jreback commented on pandas/tests/test_algos.py:
```diff
@@ -326,6 +326,78 @@ def test_factorize_na_sentinel(self, sort, na_sentinel, data, uniques):
+        codes, uniques = algos.factorize(data, dropna=dropna)
+
+        tm.assert_numpy_array_equal(uniques, expected_uniques)
+        tm.assert_numpy_array_equal(codes, expected_codes, check_dtype=False)
```
same as above

charlesdong1991 replied:
removed

jreback reviewed (changes requested):
@jbrockmendel @TomAugspurger if you'd have a look

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -191,6 +192,31 @@ method on a :func:`pandas.api.indexers.BaseIndexer` subclass that will generate
 indices used for each window during the rolling aggregation. For more details and example usage, see
 the :ref:`custom window rolling documentation <stats.custom_rolling_window>`
 
+.. _whatsnew_1000.groupby_key:
```
move to 1.1

charlesdong1991 replied:
moved

jreback reviewed (changes requested):
if you rebase and address comments can look again

jsignell commented:
Is this PR still active? I just ran into this while trying to implement a [small feature in dask](https://github.com/dask/dask/issues/5306) that this is blocking. Please let me know if you need any help @charlesdong1991

charlesdong1991 committed:
merge master and resolve conflicts

charlesdong1991 committed:
remove blank

charlesdong1991 committed:
code change on reviews

charlesdong1991 committed:
use pd.testing

charlesdong1991 committed:
linting

charlesdong1991 committed:
fixup

charlesdong1991 committed:
fixup

charlesdong1991 commented on doc/source/user_guide/groupby.rst:
```diff
@@ -199,6 +199,28 @@ For example, the groups created by ``groupby()`` below are in the order they app
    df3.groupby(['X']).get_group('B')
 
 
+.. _groupby.dropna:
```
added below

charlesdong1991 commented on doc/source/user_guide/groupby.rst:
```diff
@@ -199,6 +199,28 @@ For example, the groups created by ``groupby()`` below are in the order they app
+.. ipython:: python
+
+    df_list = [[1, 2, 3], [1, None, 4], [2, 1, 3], [1, 2, 2]]
+    df_dropna = pd.DataFrame(df_list, columns=["a", "b", "c"])
```
changed

charlesdong1991 committed:
doc

charlesdong1991 commented:
@jsignell Thanks for pushing it, and i just updated the PR to address reviews. But I think it might not be close to merge, and could take time for discussions. 

@jreback I think I addressed most of your previous comments, except https://github.com/pandas-dev/pandas/pull/30584#discussion_r363406093 and this is the reason why the index type check has to be set to False, however, I do think the result of this is correct, and might need to think about a bit how MI should deal with NA, i opened the issue before #30750 and the pd.testing here constructs MI via `from_*` and will automatically drop NA in levels, and this lead to the `inferred_type` to `string`, and I think `mixed` should be correct input here, because we do include NA in levels, and thus `inferred_type` should be `mixed`.

charlesdong1991 committed:
validation

jsignell commented:
Wow that was a quick response @charlesdong1991! I'm not in any particular rush, just wanted to bump this back up in people's todo lists.

charlesdong1991 committed:
xfail windows

charlesdong1991 committed:
rebase and resolve conflict

WillAyd commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -36,6 +36,31 @@ For example:
    ser.loc["May 2015"]
 
+
+.. _whatsnew_1000.groupby_key:
```
```suggestion
.. _whatsnew_110.groupby_key:
```

WillAyd commented on pandas/core/generic.py:
```diff
@@ -7350,6 +7350,12 @@ def clip(
+            with row/column will be dropped.
+            If False, NA values will also be treated as the key in groups
+
+            .. versionadded:: 1.0.0
```
```suggestion
            .. versionadded:: 1.1.0
```

charlesdong1991 committed:
fixup based on WA review

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into fix_issue_3729

charlesdong1991 commented:
@WillAyd thanks for your reviews, and I had changes based on your reviews, feel free to give another look!

WillAyd reviewed (changes requested):
@charlesdong1991 generally looks OK; the big thing here is simplifying / making the tests clearer

charlesdong1991 committed:
fix conflicts

charlesdong1991 committed:
reduce tests

pep8speaks commented:
Hello @charlesdong1991! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:













* In the file [`pandas/tests/groupby/test_groupby_dropna.py`](https://github.com/pandas-dev/pandas/blob/3f360a997a6e924c754906a6ca34b217bfe0c45f/pandas/tests/groupby/test_groupby_dropna.py):

> [Line 216:61](https://github.com/pandas-dev/pandas/blob/3f360a997a6e924c754906a6ca34b217bfe0c45f/pandas/tests/groupby/test_groupby_dropna.py#L216): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

charlesdong1991 commented:
@WillAyd thanks for taking a look again!

I reduce the tests quite massively, and should look simpler! thanks for the suggestion!

charlesdong1991 committed:
fix pep8

jreback commented on doc/source/user_guide/groupby.rst:
```diff
@@ -199,6 +199,27 @@ For example, the groups created by ``groupby()`` below are in the order they app
+
+The default setting of ``dropna`` argument is ``True`` which means ``NA`` are not included in group keys.
+
+.. versionadded:: 1.1.0
```
this tag goes *before* the `groupby.dropna` tag

charlesdong1991 replied:
moved above!

jreback commented on doc/source/user_guide/groupby.rst:
```diff
@@ -199,6 +199,27 @@ For example, the groups created by ``groupby()`` below are in the order they app
+    df_list = [[1, 2, 3], [1, None, 4], [2, 1, 3], [1, 2, 2]]
+    df_dropna = pd.DataFrame(df_list, columns=["a", "b", "c"])
+
+    df_dropna.groupby(by=["b"], dropna=False).sum()
```
reorg these to be

```
# say this is the default
df_dropna.groupby(by=["b"], dropna=True).sum()

# to allow NaN keys
```

but use full sentences and such

charlesdong1991 replied:
changed!

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -36,6 +36,31 @@ For example:
+We've added a ``dropna`` keyword to :meth:`DataFrame.groupby` and :meth:`Series.groupby` in order to
+allow ``NA`` values in group keys. Users can define ``dropna`` to ``False`` if they want to include
+``NA`` values in groupby keys. The default is set to ``True`` for ``dropna`` to keep backwards
+compatibility (:issue:`3729`)
```
add a ref to the new doc-section.

charlesdong1991 replied:
emm, what does this mean?

jreback replied:
you add something like: ```:ref:`cookbook<cookbook.grouping>` ``` to put in a link to the new groupby section you added about (obviously use the new ref you created)

charlesdong1991 replied:
thanks for the hint! I added one following other examples, pls let me know if it is okay now

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -36,6 +36,31 @@ For example:
+
+.. ipython:: python
+
+    df_list = [[1, 2, 3], [1, None, 4], [2, 1, 3], [1, 2, 2]]
```
use the same example as in the docs section (e.g. make the changes here as well)

charlesdong1991 replied:
yeah, the example is the same as in `groupby.rst` if it is what you meant. I also make changes as suggested above.

jreback replied:
same change here as above

jreback commented on pandas/core/groupby/groupby.py:
```diff
@@ -397,6 +398,8 @@ def __init__(
         self.observed = observed
         self.mutated = mutated
 
+        self.dropna = dropna
```
i think you added a line here?

charlesdong1991 replied:
removed!

jreback commented on pandas/core/groupby/grouper.py:
```diff
@@ -260,6 +265,8 @@ def __init__(
         self.observed = observed
         self.in_axis = in_axis
 
+        self.dropna = dropna
```
extra line is added here again?

charlesdong1991 replied:
removed!

jreback commented on pandas/tests/groupby/test_groupby_dropna.py:
```diff
@@ -0,0 +1,237 @@
+    tm.assert_frame_equal(grouped, expected, check_index_type=False)
+
+
+@pytest.mark.parametrize(
```
use the nulls_fixture & nulls_fixture2

charlesdong1991 replied:
changed!

jreback commented on pandas/tests/groupby/test_groupby_dropna.py:
```diff
@@ -0,0 +1,237 @@
+import pandas.testing as tm
+
+
+@pytest.mark.parametrize("na_value", [np.nan, None])
```
use nulls_fixture

charlesdong1991 replied:
changed!

jreback commented on pandas/tests/groupby/test_groupby_dropna.py:
```diff
@@ -0,0 +1,237 @@
+    "na_value1, na_value2",
+    [
+        (np.nan, pd.NaT),
+        (np.nan, np.nan),
```
hmm is there a reason you are mixing np.nan and pd.NaT?

charlesdong1991 replied:
emm, not really, just wanted to prove those combinations all work. Note that here cannot use `nulls_fixture` because all cases including `pd.NA` will fail.

jreback commented on pandas/tests/groupby/test_groupby_dropna.py:
```diff
@@ -0,0 +1,237 @@
+def test_groupby_dropna_datetime_like_data(
+    na_value1, na_value2, dropna, values, datetime1, datetime2
+):
+    # 3729
```
ideally check this for timedeltas / period as well

charlesdong1991 replied:
i think i did included timedeltas and period, see `datetime1`, and `datetime2`? i have renamed this function to `datetime_like` to avoid confusion, or could you please suggestion some other name if this sounds still confusing?

jreback commented on pandas/tests/test_algos.py:
```diff
@@ -326,6 +326,86 @@ def test_factorize_na_sentinel(self, sort, na_sentinel, data, uniques):
 
+    @pytest.mark.xfail(
+        compat.is_platform_windows(),
+        reason="Windows will be coelced to int32 other than int64",
```
huh? what is changed to int32? simply construct the expected with a fixed dtype, we don't want to xfail things.

charlesdong1991 replied:
emm, i got pretty weird errors before only on windows builds, and complained the `dtype` is different, while getting `int32` but expected is `int64`, so i assume something is different in windows that makes this different

jreback reviewed (changes requested):
looks pretty good. we *always* want to check the index type of the return, construct the expected so that it matches

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into fix_issue_3729

charlesdong1991 committed:
rebase and docs fixes

charlesdong1991 committed:
fixup doc

charlesdong1991 committed:
remove inferred type

charlesdong1991 committed:
better comment

charlesdong1991 committed:
remove xfail

charlesdong1991 committed:
use fixture

charlesdong1991 committed:
coelse type for windows build

charlesdong1991 committed:
fixup

charlesdong1991 committed:
fixup

charlesdong1991 commented:
Hi, @WillAyd @jreback 

I feel that we see the light in the end of tunnel for this PR 😄 quick summary to address all your comments:
1. Removed `check_index_type=False`! which i think was the biggest issue before, the reason it was there was when constructing MI for `expected`, if `NaN` in there, MI will automatically drop it from levels, so I need to manually set level for `expected` to include `NaN` if `dropna=False`
2. Rephrase/Reorg docs in `groupby.rst` and `1.1.0rst`, but I think there are still a couple comments about the `ref` is not clear to me, feel free to complement!
3. Simplified the tests significantly: basically compress all 3 similar tests for datetime-alike into one test with parametrization, and replace `na_values` in parametrization with existing `nulls_fixture` and `unique_nulls_fixture` from `conftest.py`
4. Remove xfail for windows machine builds in `test_algo.py`, the reason it was there was integer array has by default dtype of `int32` in windows i assume, so there is `dtype` conflict, therefore, I typed fixed dtype for it to avoid using xfail.

Feel free to take a look again, and hopefully lets get it merged soon!! ^^

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into fix_issue_3729

charlesdong1991 commented:
any followups on this one? @jreback @WillAyd ^^

jreback reviewed (changes requested):
looks really good, just a couple of doc comments. ping on green.

jreback commented:
@jorisvandenbossche @TomAugspurger @jbrockmendel if you can have a quick skim.

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into fix_issue_3729

charlesdong1991 committed:
Doc fixup

charlesdong1991 commented on doc/source/user_guide/groupby.rst:
```diff
@@ -199,6 +199,29 @@ For example, the groups created by ``groupby()`` below are in the order they app
+
+    df_list = [[1, 2, 3], [1, None, 4], [2, 1, 3], [1, 2, 2]]
+    df_dropna = pd.DataFrame(df_list, columns=["a", "b", "c"])
+
```
i think you meant `df_dropna`, i showed it and put groupbys to another block

charlesdong1991 commented:
ping @jreback while waiting for other reviews

p.s. somehow, there are less-than-usual CI checks after a rebase (i expect at least 13 or so, but see only 6, seems a lot of checks on other OS or python version are not triggered), not sure how/why, and pls let me know if another rebase is needed to see if the change is green on all.

jreback commented:
> ping @jreback while waiting for other reviews
> 
> p.s. somehow, there are less-than-usual CI checks after a rebase (i expect at least 13 or so, but see only 6, seems a lot of checks on other OS or python version are not triggered), not sure how/why, and pls let me know if another rebase is needed to see if the change is green on all.

hmm, try rebasing again, this happened once before then resolved itself.

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into fix_issue_3729

charlesdong1991 commented:
emm, weird, indeed seems it resolves itself. @jreback

charlesdong1991 committed:
rebase and resolve conflict

charlesdong1991 commented:
ping for reviews ^^

@jreback @jbrockmendel @TomAugspurger @jorisvandenbossche

TomAugspurger commented:
Don't wait around on me.

On Tue, Apr 28, 2020 at 5:21 AM Kaiqi Dong <notifications@github.com> wrote:

> ping for reviews ^^
>
> @jreback <https://github.com/jreback> @jbrockmendel
> <https://github.com/jbrockmendel> @TomAugspurger
> <https://github.com/TomAugspurger> @jorisvandenbossche
> <https://github.com/jorisvandenbossche>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/30584#issuecomment-620517194>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKAOIQFPAQ3UJ52HM43DWLRO2U4PANCNFSM4KBWP3MQ>
> .
>

charlesdong1991 commented:
maybe @WillAyd  @jbrockmendel could take a look for some feedbacks?

WillAyd reviewed (approved):
I think this is a nice change. @jreback

jreback commented:
@charlesdong1991 can you merge master and ping on green.

charlesdong1991 committed:
Merge remote-tracking branch 'upstream/master' into fix_issue_3729

charlesdong1991 committed:
try merge master again

charlesdong1991 commented:
ping @jreback @WillAyd

jreback commented:
thanks @charlesdong1991 very nice! this has been a long time requested feature!

senegrom commented:
stupid question: how does that feature propagate into the standard pandas branches? any timeframe?

jsignell commented:
My understanding is this is slated for inclusion in pandas 1.1.0, but I'm not sure when that is planned for.
